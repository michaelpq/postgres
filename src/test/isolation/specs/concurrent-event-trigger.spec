# Test for interactions with DDL commands manipulating event trigger
# objects.
# The following set of commands can interact in concurrency:
# - ALTER EVENT TRIGGER ENABLE
# - ALTER EVENT TRIGGER DISABLE
# - ALTER EVENT TRIGGER OWNER TO

setup
{
  CREATE OR REPLACE FUNCTION notice_any_command()
    RETURNS event_trigger
    LANGUAGE plpgsql
    AS $$
      BEGIN
      RAISE NOTICE 'command % is run', tg_tag;
      END;
    $$;
  CREATE EVENT TRIGGER notice_ddl ON ddl_command_start
     EXECUTE PROCEDURE notice_any_command();
}

teardown
{
  DROP EVENT TRIGGER notice_ddl;
  DROP FUNCTION notice_any_command();
}

session "s1"
step "s1_begin"		{ BEGIN; }
step "s1_disable"	{ ALTER EVENT TRIGGER notice_ddl DISABLE; }
step "s1_enable"	{ ALTER EVENT TRIGGER notice_ddl ENABLE REPLICA; }
step "s1_owner"		{ ALTER EVENT TRIGGER notice_ddl OWNER TO CURRENT_USER; }
step "s1_commit"	{ COMMIT; }

session "s2"
step "s2_begin"		{ BEGIN; }
step "s2_disable"	{ ALTER EVENT TRIGGER notice_ddl DISABLE; }
step "s2_enable"	{ ALTER EVENT TRIGGER notice_ddl ENABLE REPLICA; }
step "s2_owner"		{ ALTER EVENT TRIGGER notice_ddl OWNER TO CURRENT_USER; }
step "s2_commit"	{ COMMIT; }

permutation "s1_begin" "s1_disable" "s2_begin" "s2_disable"
  "s1_commit" "s2_commit"
permutation "s1_begin" "s1_disable" "s2_begin" "s2_enable"
  "s1_commit" "s2_commit"
permutation "s1_begin" "s1_disable" "s2_begin" "s2_owner"
  "s1_commit" "s2_commit"
permutation "s1_begin" "s1_enable" "s2_begin" "s2_enable"
  "s1_commit" "s2_commit"
permutation "s1_begin" "s1_enable" "s2_begin" "s2_owner"
  "s1_commit" "s2_commit"
permutation "s1_begin" "s1_owner" "s2_begin" "s2_owner"
  "s1_commit" "s2_commit"
