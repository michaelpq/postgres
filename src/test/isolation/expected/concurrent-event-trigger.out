Parsed test spec with 2 sessions

starting permutation: s1_begin s1_disable s2_begin s2_disable s1_commit s2_commit
step s1_begin: BEGIN;
step s1_disable: ALTER EVENT TRIGGER notice_ddl DISABLE;
step s2_begin: BEGIN;
step s2_disable: ALTER EVENT TRIGGER notice_ddl DISABLE; <waiting ...>
step s1_commit: COMMIT;
step s2_disable: <... completed>
error in steps s1_commit s2_disable: ERROR:  tuple concurrently updated
step s2_commit: COMMIT;

starting permutation: s1_begin s1_disable s2_begin s2_enable s1_commit s2_commit
step s1_begin: BEGIN;
step s1_disable: ALTER EVENT TRIGGER notice_ddl DISABLE;
step s2_begin: BEGIN;
step s2_enable: ALTER EVENT TRIGGER notice_ddl ENABLE REPLICA; <waiting ...>
step s1_commit: COMMIT;
step s2_enable: <... completed>
error in steps s1_commit s2_enable: ERROR:  tuple concurrently updated
step s2_commit: COMMIT;

starting permutation: s1_begin s1_disable s2_begin s2_owner s1_commit s2_commit
step s1_begin: BEGIN;
step s1_disable: ALTER EVENT TRIGGER notice_ddl DISABLE;
step s2_begin: BEGIN;
step s2_owner: ALTER EVENT TRIGGER notice_ddl OWNER TO CURRENT_USER;
step s1_commit: COMMIT;
step s2_commit: COMMIT;

starting permutation: s1_begin s1_enable s2_begin s2_enable s1_commit s2_commit
step s1_begin: BEGIN;
step s1_enable: ALTER EVENT TRIGGER notice_ddl ENABLE REPLICA;
step s2_begin: BEGIN;
step s2_enable: ALTER EVENT TRIGGER notice_ddl ENABLE REPLICA; <waiting ...>
step s1_commit: COMMIT;
step s2_enable: <... completed>
error in steps s1_commit s2_enable: ERROR:  tuple concurrently updated
step s2_commit: COMMIT;

starting permutation: s1_begin s1_enable s2_begin s2_owner s1_commit s2_commit
step s1_begin: BEGIN;
step s1_enable: ALTER EVENT TRIGGER notice_ddl ENABLE REPLICA;
step s2_begin: BEGIN;
step s2_owner: ALTER EVENT TRIGGER notice_ddl OWNER TO CURRENT_USER;
step s1_commit: COMMIT;
step s2_commit: COMMIT;

starting permutation: s1_begin s1_owner s2_begin s2_owner s1_commit s2_commit
step s1_begin: BEGIN;
step s1_owner: ALTER EVENT TRIGGER notice_ddl OWNER TO CURRENT_USER;
step s2_begin: BEGIN;
step s2_owner: ALTER EVENT TRIGGER notice_ddl OWNER TO CURRENT_USER;
step s1_commit: COMMIT;
step s2_commit: COMMIT;
